File: Src/CompareEngines/ExistenceCompare.cpp
/**
 * @file  ExistenceCompare.cpp
 *
 * @brief Implementation file for ExistenceCompare
 */

#include "pch.h"
#include "ExistenceCompare.h"
#include "DiffContext.h"
#include "DiffItem.h"

using Poco::Timestamp;

namespace CompareEngines
{

ExistenceCompare::ExistenceCompare(CDiffContext& ctxt)
	: m_nfiles(ctxt.GetCompareDirs())
{
}

ExistenceCompare::~ExistenceCompare() = default;

/**
 * @brief Compare specified files by their existence
 * @param [in,out] di Diffitem info. Results are written to di.diffcode.
 * @return DIFFCODE
 */
int ExistenceCompare::CompareFiles(DIFFITEM& di) const
{
	unsigned code = DIFFCODE::SAME;
	if (di.diffcode.exists(0) != di.diffcode.exists(1) ||
		(m_nfiles > 2 && di.diffcode.exists(0) != di.diffcode.exists(2)))
	{
		code = DIFFCODE::DIFF;
	}

	if (m_nfiles > 2 && (code & DIFFCODE::COMPAREFLAGS) == DIFFCODE::DIFF)
	{
		if (di.diffcode.exists(1) == di.diffcode.exists(2))
			code |= DIFFCODE::DIFF1STONLY;
		else if (di.diffcode.exists(0) == di.diffcode.exists(2))
			code |= DIFFCODE::DIFF2NDONLY;
		else if (di.diffcode.exists(0) == di.diffcode.exists(1))
			code |= DIFFCODE::DIFF3RDONLY;
	}

	di.diffcode.diffcode &= ~(DIFFCODE::TEXTFLAGS | DIFFCODE::TYPEFLAGS | DIFFCODE::COMPAREFLAGS | DIFFCODE::COMPAREFLAGS3WAY);
	di.diffcode.diffcode |= DIFFCODE::FILE | code;
	return code;
}

} // namespace CompareEngines

File: winmerge-core/src/main/java/org/winmerge/core/compare/engines/ExistenceCompare.java
package org.winmerge.core.compare.engines;

import org.winmerge.core.compare.DiffCode;

public final class ExistenceCompare implements CompareEngine {
    @Override
    public CompareEngineResult compare(CompareEngineContext context) {
        context.refreshFileInfos();
        int files = context.getFileCount();
        int code = DiffCode.SAME;
        if (context.exists(0) != context.exists(1)
                || (files > 2 && context.exists(0) != context.exists(2))) {
            code = DiffCode.DIFF;
        }

        if (files > 2 && (code & DiffCode.COMPAREFLAGS) == DiffCode.DIFF) {
            if (context.exists(1) == context.exists(2)) {
                code |= DiffCode.DIFF1STONLY;
            } else if (context.exists(0) == context.exists(2)) {
                code |= DiffCode.DIFF2NDONLY;
            } else if (context.exists(0) == context.exists(1)) {
                code |= DiffCode.DIFF3RDONLY;
            }
        }

        return CompareEngineResult.withUnknownCounts(DiffCode.FILE | code);
    }
}
